rflx/expression.py

```diff
@@ -2009,26 +2069,95 @@ class MessageAggregate(Expr):
     def __init__(
         self,
         name: StrID,
-        data: Mapping[StrID, Expr],
+        field_values: Mapping[StrID, Expr],
         type_: rty.Type = rty.Undefined(),
         location: Location = None,
     ) -> None:
         super().__init__(type_, location)
         self.name = ID(name)
-        self.data = {ID(k): v for k, v in data.items()}
+        self.field_values = {ID(k): v for k, v in field_values.items()}
 
     def _update_str(self) -> None:
-        data = (
-            ", ".join(f"{k} => {self.data[k]}" for k in self.data) if self.data else "null message"
+        field_values = (
+            ", ".join(f"{k} => {self.field_values[k]}" for k in self.field_values)
+            if self.field_values
+            else "null message"
         )
-        self._str = intern(f"{self.name}'({data})")
+        self._str = intern(f"{self.name}'({field_values})")
+
+    def _check_type_subexpr(self) -> RecordFluxError:
+        if not isinstance(self.type_, rty.Message):
+            error = RecordFluxError()
+
+            for d in self.field_values.values():
+                error += d.check_type_instance(rty.Any)
+
+            return error
+
+        error = RecordFluxError()
+        field_combinations = set(self.type_.field_combinations)
+
+        for i, (field, expr) in enumerate(self.field_values.items()):
+            if str(field) not in self.type_.fields:
+                error.append(
+                    f'invalid field "{field}" for {self.type_}',
+                    Subsystem.MODEL,
+                    Severity.ERROR,
+                    field.location,
+                )
+                error.extend(_similar_field_names(str(field), self.type_.fields, field.location))
+                continue
+
+            field_type = self.type_.field_types[str(field)]
+
+            if field_type == rty.OPAQUE:
+                for refinement_field, refinement_type in self.type_.refinements:
+                    if refinement_field == str(field) and expr.type_.is_compatible(refinement_type):
+                        break
+                else:
+                    error += expr.check_type(field_type)
```

Comment:
I think this could be:
```Python
if any(refinement_field == str(field) and expr.type_.is_compatible(refinement_type) for refinedment_field, refinement_type in self.type_.refinements):
    error += expr.check_type(field_type)
```

Reply:
It must be `not any`, but I like the idea. Changed in 2317ee40cdae0e927082bf4b98ad1cb2c6cc922c.

Reply:
Instead of `not any` you also could use `all`.

Reply:
I also thought about that, but I think that the `not any` variant is easier to understand.

Reply:
Tbh I think it's the other way. I think it's easier to understand that all element must match the expression than not at least one element must match it. @senier what's your opinion on this?

Reply:
The crucial point for me is that the negated expression is harder to understand, but maybe this is just a matter of personal preference:


```Python
                if not any(
                    f == str(field) and expr.type_.is_compatible(t)
                    for f, t in self.type_.refinements 
                ):
```

```Python
                if all(
                    f != str(field) or not expr.type_.is_compatible(t)
                    for f, t in self.type_.refinements 
                ):
```

Reply:
It's really a matter of taste. I'm leaning towards the `not any` variant, as I find "there is no matching and compatible refinement" which it translates to more suitable for understanding the cause of the error.